### PR TITLE
[Typechecker] Allow initialising a variable with a function with the same name as the variable

### PR DESCRIPTION
**Pull request by theblixguy**
Previously, we would reject it in the parser. So, something like:

```swift
func addOne(to number: Int) -> Int { return number + 1 }
var addOne = addOne(to: 5)
```

would trigger an error diagnostic: `Variable used in its own initial value`.

This is because the parser would think that we're trying to self-reference the variable, even in cases where we weren't (like above). 

Now, the parser simply ignores the possibility of self-reference and passes an `UnresolvedDeclRefExpr` up to the type checker. The type checker is then responsible for checking self-reference and generating the diagnostics. If we're not trying to self-reference, then the `UnresolvedDeclRefExpr` gets resolved as expected.

Resolves [SR-10295](https://bugs.swift.org/browse/SR-10295), [SR-10233](https://bugs.swift.org/browse/SR-10233), [SR-5916](https://bugs.swift.org/browse/SR-5916), [SR-3040](https://bugs.swift.org/browse/SR-3040) & [SR-1846](https://bugs.swift.org/browse/SR-1846)
Resolves rdar://problem/22374793.
Resolves rdar://problem/45572606.

### CODE REVIEW

**Comment by theblixguy** [comment]
There is one scenario which I discovered where it leads to a crash:

```
struct Person {
    let firstName: String
    let lastName: String
}

class Klass {
  func name(for person: Person) -> String {
  return person.firstName + " " + person.lastName
}

func foo() {
   let helloWorld = Person(firstName: "Hello", lastName: "World")
   let name = name(for: helloWorld)
  }
}
```

The crash happens when `resolveOverload()` calls `getTypeOfReference()` which hits an assertion: `Assertion failed: (func->isOperator() && "Lookup should only find operators")`. I am confused as to why this assertion is being hit, even though I am returning the correct `DeclRefExpr` for `name`.

This doesn't happen when you move the code outside of the function or make it global.

cc @jrose-apple @xedin

File: test/expr/expressions.swift [hunk 1]
```diff
@@ -244,13 +244,12 @@ func test_lambda() {
...
   // A recursive lambda.
-  // FIXME: This should definitely be accepted.
-  var fib = { (n: Int) -> Int in
+  var fib: (Int) -> Int = { (n: Int) -> Int in
     if (n < 2) {
       return n
     }
     
-    return fib(n-1)+fib(n-2) // expected-error 2 {{variable used within its own initial value}}
+    return fib(n-1)+fib(n-2)
```
Comment (jrose-apple) [inline test/expr/expressions.swift]: This has to continue diagnosing; the variable can't be captured like this.

**Review by jrose-apple — COMMENTED** [review]
I don't think this is the correct approach for this change, if we even agree it's the right thing to do. Instead, we should be modifying unqualified lookup to ignore the variable within the initializer body, or to allow overloading with it, or something. Doing ad-hoc recovery by looking at the results of lookup means losing some of the original context of how the lookup was performed.

**Comment by jrose-apple** [comment]
In your crash example, I think we want a MemberRefExpr with an implicit `self`, not a DeclRefExpr.

File: test/expr/expressions.swift (hunk 1 above)
Comment (theblixguy) [reply test/expr/expressions.swift]: Ah ok, I saw the fix-me above, so I updated the code. Without the explicit type annotation, it gets diagnosed. Perhaps I should look into the bodies of closures as well which have more than 1 expression.

**Comment by theblixguy** [comment]
Thanks @jrose-apple, I'll explore other approaches to this. UnqualifiedLookup does return the correct function in some cases (and I _think_ we end up creating an `OverloadedDeclRefExpr`), so the additional code in the bottom doesn't get triggered, although yes in some cases, it doesn't quite work as expected so we have to recover in a ad-hoc way. Let me look into another way of fixing it.

**Comment by theblixguy** [comment]
@jrose-apple I have added a new flag to UnqualifiedLookup to ignore local variables. This seems to work without needing ad-hoc recovery. Could you take a look at my recent changes? Thank you!

File: include/swift/AST/NameLookup.h [hunk 2]
```diff
@@ -101,6 +101,9 @@ class UnqualifiedLookup {
     /// This lookup should include results from outside the innermost scope with
     /// results.
     IncludeOuterResults = 0x10,
+    /// Whether to ignore local variables inside bodies of functions, closures
+    /// and top level code.
+    IgnoreLocalVariables = 0x50,
```
Comment (jrose-apple) [inline include/swift/AST/NameLookup.h]: This isn't a power of 2. Is it supposed to include IncludeOuterResults?

File: lib/Sema/TypeChecker.h [hunk 3]
```diff
@@ -305,6 +305,9 @@ enum class NameLookupFlags {
   IncludeOuterResults = 0x20,
   /// Whether to consider synonyms declared through @_implements().
   IncludeAttributeImplements = 0x40,
+  /// Whether to ignore local variables inside bodies of functions, closures
+  /// and top level code.
+  IgnoreLocalVariables = 0x50,
```
Comment (jrose-apple) [inline lib/Sema/TypeChecker.h]: Same here.

File: include/swift/AST/NameLookup.h (hunk 2 above)
Comment (theblixguy) [reply include/swift/AST/NameLookup.h]: Sorry, my bad, I completely ignored the fact that the values above are hex 🤦‍♂️. Fixed now!

File: lib/Sema/TypeChecker.h (hunk 3 above)
Comment (theblixguy) [reply lib/Sema/TypeChecker.h]: Fixed!

**Comment by theblixguy** [comment]
Does this look good now? All the tests pass locally without regressions. Might be worth running source compat tests just to be sure.

**Comment by jrose-apple** [comment]
I'm sorry, I don't have time to look at this one right now, and I still think it deserves discussion given our larger worries about shadowing.

@swift-ci Please test source compatibility

**Comment by theblixguy** [comment]
No problem :) is there some one else who can help with the review or discussion? Thanks!

**Comment by jrose-apple** [comment]
Honestly, I think Doug would want to be involved, and he's pretty bogged down with the property delegates pitch at the moment. You could bring this up on the forums regardless, though.

**Comment by theblixguy** [comment]
Ah ok, I understand he's busy. Is it possible for you to request a review from him so he can take a look when he's got some free time? If not then it's fine, I'll ping again in a week. Thanks!

**Comment by jrose-apple** [comment]
Sure, let's put him on the list.

**Comment by theblixguy** [comment]
Been a week so pinging @DougGregor again. Does it look good to you, Doug?

**Comment by theblixguy** [comment]
Ping. @DougGregor does this look good to you? Thank you!

**Comment by theblixguy** [comment]
Ping @DougGregor, are you free to take a look at this? Thank you! If not, then is there anyone else who can take a look @jrose-apple? Maybe Joe or John? It's been a month now so it would be nice if we can get this reviewed 😄

**Comment by slavapestov** [comment]
I think it's worth rethinking some of the surrounding code. The core issue here is that we're not forming an overload set that contains both an unqualified and self-qualified reference from different scopes. We should do that instead of adding a new name lookup flag and having to plumb it through. In fact name lookup already has support for this, but it's only enabled in a narrow case.

What happens if you change shouldConsiderOuterResultsFor() to always return true?

**Comment by theblixguy** [comment]
@slavapestov We would end up emitting the "ambiguous use of ..." diagnostic.

**Comment by slavapestov** [comment]
@theblixguy Are you talking about this part?

```
  // FIXME: If we reach this point, the program we're being handed is likely
  // very broken, but it's still conceivable that this may happen due to
  // invalid shadowed declarations.
  //
  // Make sure we emit a diagnostic, since returning an ErrorExpr without
  // producing one will break things downstream.
  diagnose(Loc, diag::ambiguous_decl_ref, Name);
```

I guess the problem is we don't have a way to represent an overload set that contains both self-qualified and unqualified members (eg, local variables).

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 4]
```diff
@@ -323,6 +325,14 @@ LookupResult TypeChecker::lookupUnqualified(DeclContext *dc, DeclName name,
       assert(foundInType && "bogus base declaration?");
     }
 
+    // This is a hack to filter out top level variables. If this flag is set
+    // then local variables are ignored, but for some reason, top level
+    // variables do not get ignored, so ignore them here.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckNameLookup.cpp]: Top-level variables are member-qualified by the module, whereas local variables are unqualified.

**Comment by slavapestov** [comment]
I think removing the circularity check from the parser is the right way to go, but we should instead handle this via the overload mechanism instead of further complicating name lookup and resolveDeclRefExpr().

I found another hack that's similar -- look at how resolveDeclRefExpr() does this:

```
while (localDeclAfterUse) {
...
      Lookup.shiftDownResults();
...
}
```

I think we should try to combine your new check with the existing 'local decl after use' logic. We need to be able to form overloads with both qualified and unqualified bases, and we need to filter out overloads that match certain criteria (implementing the circularity and use-before-declaration checks).

File: lib/Parse/ParseExpr.cpp [hunk 5]
```diff
@@ -2288,8 +2288,13 @@ Expr *Parser::parseExprIdentifier() {
     } else {
       for (auto activeVar : DisabledVars) {
         if (activeVar->getFullName() == name) {
-          diagnose(loc.getBaseNameLoc(), DisabledVarReason);
-          return new (Context) ErrorExpr(loc.getSourceRange());
+          if (DisabledVarReason.ID == diag::var_init_self_referential.ID) {
```
Comment (slavapestov) [inline lib/Parse/ParseExpr.cpp]: How about not disabling the var at all in this case then?

**Comment by slavapestov** [comment]
@theblixguy What should this example do?

```
func outer() {}

func blah() {
  var outer = 10
  var a = outer()
}
```

Right now, we can't form an overload set for the call 'outer()' at all, because there's no circularity so `outer` doesn't get disabled -- so the parser forms a DeclRefExpr and resolveDeclRefExpr() is not called.

**Comment by theblixguy** [comment]
That would currently fail with `cannot call value of non-function type 'Int'`.

File: lib/Parse/ParseExpr.cpp (hunk 5 above)
Comment (theblixguy) [reply lib/Parse/ParseExpr.cpp]: Hmm yeah that sounds like a better idea than this if check!

**Comment by theblixguy** [comment]
> I think removing the circularity check from the parser is the right way to go, but we should instead handle this via the overload mechanism instead of further complicating name lookup and resolveDeclRefExpr().

You mean inside `resolveDeclRefExpr()`, we return an `OverloadedDeclRefExpr` which includes the both the variable and function decl so it can be resolved higher up (in the constraint system) instead of doing name lookup to resolve the UDRE?

> 
> I found another hack that's similar -- look at how resolveDeclRefExpr() does this:
> 
> ```
> while (localDeclAfterUse) {
> ...
>       Lookup.shiftDownResults();
> ...
> }
> ```
> 
> I think we should try to combine your new check with the existing 'local decl after use' logic. We need to be able to form overloads with both qualified and unqualified bases, and we need to filter out overloads that match certain criteria (implementing the circularity and use-before-declaration checks).

Sorry, I didn't fully understand the "combine your new check with the existing 'local decl after use' logic" bit.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**